### PR TITLE
Calculate assignment without abstains (if one candidate) (issue #2182)

### PR DESCRIPTION
**Pull request by matakuka**
Follow-up for issue #2182 and PR #2192:

- Added a new variable to each assignment poll: `without_abstain`.
- if this variable is set true and there is only one candidate in the assignment options, the percentages will be returned with the base <s>(votesvalid or votescast) reduced by amount of abstains.</s> `votes.yes + votes.no` .
- The option may only be toggled if there is only one candidate (button on polll result page).
- calculation is effective on projector and result view, but not (yet) on pdf creation (this is still created in python)

To discuss:  
<s>- Is `base = Votescast - abstains` a valid calculation?</s>
- How to handle polls with two candidates and two posts?
- How to handle poll #n when there is only one candidate left, but assignment.options has had more?

To do:
- Avoid reloading of whole poll, reload numbers and percentages instead.

### CODE REVIEW

File: openslides/assignments/models.py [hunk 1]
```diff
@@ -366,6 +366,7 @@ class AssignmentPoll(RESTModelMixin, CollectDefaultVotesMixin,
         related_name='polls')
     yesnoabstain = models.BooleanField(default=False)
     yesno = models.BooleanField(default=False)
+    without_abstains = models.BooleanField(default=False)
```
Comment (normanjaeckel) [inline openslides/assignments/models.py]: Sorry. I though we take an aproach without new database values. The percent base is just calculated on the fly if config and situation matches.

**Comment by normanjaeckel** [comment]
I would like to have a non database solution, means the percent base should be calculated on the fly.

**Comment by matakuka** [comment]
I reinserted it because @emanuelschuetze wanted "if the option is set, it should be set for all clients." Such a distribution of requirements doesn't work without database. I can do either, but as it seems, some discussion is needed.

**Comment by normanjaeckel** [comment]
I do not understand what "for all clients" mean. If we use a config option (as in motion) then all clients can calculate the percent base according to the same algorithm.

**Comment by matakuka** [comment]
What was meant: 
Once this option is set (by e.g. admin), in every calculation it should show with the new percent base. So that every display and every pdf printout shows the same value.

**Comment by matakuka** [comment]
Another rework (assignment and motions) after discussion.

server side:
- poll option config values changed to "YES_NO, "YES_NO_ABSTAIN", "DISABLED" (removed "all votes (including invalid)")
- poll.yesno (bool) and poll.yesnoabstain (bool) were merged to poll.pollmethod (string) to avoid situations where both are true.
- assignment pdf percent values are disabled for now. We will move to js pdf creation soon.

client side:
- percentage calculations are now based on sums of 'yes+no', 'yes+no+abstains' or 'sum of yes of all candidates', depending on the setting. If one of them is negative, percentages should not be calculated.
- (in)valid votes are renamed to "(in)valid ballot" and don't show "100%" anymore. To discuss whether this is the correct name. We want to distinguish the number of papers from the votes (as the number is not always the sum of the votes)

For the percentages, we are now independent from the (optional) entries for votescast, votesvalid, votesinvalid.
@emanuelschuetze ready for testing

**Comment by matakuka** [comment]
rebased on master, ready for testing @emanuelschuetze

File: openslides/assignments/static/templates/assignments/assignment-detail.html [hunk 2]
```diff
@@ -177,53 +177,51 @@ <h3 translate>Election result</h3>
...
-          <td ng-if="poll.has_votes" ng-init="vote = poll.getVote(poll.votescast)">
-            {{ vote.value }} {{ vote.percent }}
-      </table>
+          <!-- total votes (valid/invalid/casts) -->
+          <tr>
+            <td>
+            <td>
+              <translate>Valid ballots</translate>
+            <td ng-if="poll.has_votes" ng-init="vote = poll.getVote(poll.votesvalid)">
+              {{ vote.value }}
```
Comment (emanuelschuetze) [inline openslides/assignments/static/templates/assignments/assignment-detail.html]: The entered values of votesvalid, votesinvalid and votescast are empty in this template... I can't see why...

**Comment by matakuka** [comment]
Other TODOs after discussion:

- [ ] rename the current config so that it is clear that the count is *per candidate*
- [ ] reintroduce the option to base the 100% on *votesvalid*
- [ ] reintroduce a calculation that shows how *votesvalid* compare to *votescast*.
- [ ] find a short, catchy wording for  "if several candidates && option != (100%= all votesvalid), then 100% will be sum of Yes of all candidates"

**Comment by matakuka** [comment]
all TODOs are done. Some (especially formulations) might depend on taste. @emanuelschuetze

**Comment by emanuelschuetze** [comment]
Retested. Some minor remarks:
* Bug: valid/invalid/votescast are still empty in pdf
* rename "total ballots" -> "casted ballots"
* remove percentages for invalid/total ballots if config option "valid votes" is selected.
* remove the config option "all ballots" (because not practically relevant that invalid ballots contains to 100% basis)
* Bug if you use Yes-election (e.g. 3 candidates for 2 posts) + config "YNA": 100% basis has to be the sum of all votes of all candidates! Curently each candidate gets 100%.
* Wording improvements for config options, use the follwing strings:
 1. Yes/No/Abstain per candidate
 1. Yes/No per candidate
 1. All valid ballots
 1. Disabled (no percents)
* Update help text, suggestion:
"For Yes/No/Abstain and Yes/No the 100% basis depends on the election method: If 'one option per candidate' is selected, the sum of all votes of all candidates is 100%. Otherwise the sum of all votes per candidate is 100%."

**Comment by emanuelschuetze** [comment]
Retested:
> * Bug: valid/invalid/votescast are still empty in pdf

Now, there are no results in PDF.

> * rename "total ballots" -> "casted ballots"

in assignment poll form is still "total ballots".

> * remove percentages for invalid/total ballots if config option "valid votes" is selected.

ok.

> * remove the config option "all ballots" (because not practically relevant that invalid ballots contains to 100% basis)

as discussed, we keep this option because some organisations need it.

> * Bug if you use Yes-election (e.g. 3 candidates for 2 posts) + config "YNA": 100% basis has to be 
> the sum of all votes of all candidates! Curently each candidate gets 100%.

Ok now.

> * Wording improvements for config options, use the follwing strings: [..]

Ok.

> * Update help text, suggestion:

Ok. In manual we will explain it more verbose...


Summary: 2 issues are still open! Rest looks good.

**Comment by emanuelschuetze** [comment]
> Now, there are no results in PDF.

My fault: I forgot to publish the results. Now all votes are listed in PDF result table (without percents which are readded if pdf is converted to pdfmake in the next weeks)

**Comment by emanuelschuetze** [comment]
@matakuka Still open: Rename "total ballots" to "casted ballots".
@normanjaeckel Ready for review.

**Comment by emanuelschuetze** [comment]
@matakuka One last thing:
* in motion detail view: remove percent for invalid votes if option 'all valid votes' is selected.

**Comment by emanuelschuetze** [comment]
Retested:
All done now. Thanks.

@normanjaeckel Ready for your review.

**Comment by matakuka** [comment]
(just a rebase on top of current master)

**Comment by normanjaeckel** [comment]
Please allow maintainer edit (checkbox in the right sidebar).

File: openslides/assignments/static/js/assignments/site.js [hunk 3]
```diff
@@ -764,8 +765,10 @@ angular.module('OpenSlidesApp.assignments.site', ['OpenSlidesApp.assignments'])
         gettext('Elections');
         gettext('Ballot and ballot papers');
         gettext('The 100 % base of an election result consists of');
+        gettext('If you chose sums per candidate, and there are more candidates than open posts,' +
```
Comment (normanjaeckel) [inline openslides/assignments/static/js/assignments/site.js]: This string is out of date.

File: openslides/assignments/static/js/assignments/site.js [hunk 4]
```diff
@@ -764,8 +765,10 @@ angular.module('OpenSlidesApp.assignments.site', ['OpenSlidesApp.assignments'])
...
+                'the base will be the sum of votes for all candidates, in order to keep the results comparable.');
         gettext('All valid votes (Yes/No/Abstain)');
-        gettext('All votes cast (including invalid votes)');
+        gettext('Yes and No Values');
```
Comment (normanjaeckel) [inline openslides/assignments/static/js/assignments/site.js]: Where is this string from?

File: openslides/motions/static/js/motions/site.js [hunk 5]
```diff
@@ -1883,7 +1882,7 @@ angular.module('OpenSlidesApp.motions.site', ['OpenSlidesApp.motions', 'OpenSlid
         gettext('Voting and ballot papers');
         gettext('The 100 % base of a voting result consists of');
         gettext('All valid votes (Yes/No/Abstain)');
-        gettext('All votes cast (including invalid votes)');
+        gettext('Yes and No Values');
```
Comment (normanjaeckel) [inline openslides/motions/static/js/motions/site.js]: Where is this string from?

File: openslides/poll/models.py [hunk 6]
```diff
@@ -95,36 +90,6 @@ def get_percent_base_choice(self):
         """
```
Comment (normanjaeckel) [inline openslides/poll/models.py]: The docstring is out of date.

File: openslides/assignments/config_variables.py [hunk 7]
```diff
@@ -28,10 +27,19 @@ def get_config_variables():
...
         input_type='choice',
         label='The 100 % base of an election result consists of',
-        choices=PERCENT_BASE_CHOICES,
+        choices=(
+            {'value': 'YES_NO_ABSTAIN', 'display_name': 'Yes/No/Abstain per candidate'},
+            {'value': 'YES_NO', 'display_name': 'Yes/No per candidate'},
+            {'value': 'VALID', 'display_name': 'All valid ballots'},
+            {'value': 'CAST', 'display_name': 'All casted ballots'},
+            {'value': 'DISABLED', 'display_name': 'Disabled (no percents)'},
+            ),
```
Comment (normanjaeckel) [inline openslides/assignments/config_variables.py]: Change indention

File: openslides/assignments/config_variables.py [hunk 8]
```diff
@@ -28,10 +27,19 @@ def get_config_variables():
...
+        help_text="For Yes/No/Abstain and Yes/No the 100% basis depends on the election method: If there are " +
```
Comment (normanjaeckel) [inline openslides/assignments/config_variables.py]: Type basis -> base. Add whitespace between 100 and %.

**Comment by emanuelschuetze** [comment]
@matakuka As discussed: wait until #2379 is merged and then rebase and use the same (generic) logic as current list of speaker slide (implemented in multiprojector pr)

**Comment by emanuelschuetze** [comment]
Ready for review and merge.

**Comment by normanjaeckel** [comment]
Please add CHANGELOG entry and rebase.

**Comment by emanuelschuetze** [comment]
@normanjaeckel ping, ready for merge

**Comment by normanjaeckel** [comment]
Hey my commits/changes were lost. I do not want to do all the work again and again.

**Comment by normanjaeckel** [comment]
So, the work is now done again. Please read carefully messages and commit list before you push force on the branch.

**Comment by matakuka** [comment]
I'm sorry. I'm not used to having others directly committing into my branch (and thus pulling my own branch before a rebase).

**Comment by normanjaeckel** [comment]
No problem. This is a new GitHub feature. :-)

**Pull request merged**